### ISSUE TITLE
[operator] need to provide schema definitions to CRDs

### ISSUE THREAD

**Issue by jmazzitelli** [kiali/kiali#1757]
Kubernetes v1.16 introduces api version `apiextensions.k8s.io/v1` of CRDs. The old `v1beta1` is deprecated.

The big change here is that our CRDs ([Kiali](https://github.com/kiali/kiali/blob/v1.5/operator/deploy/crd.yaml#L19-L40) and [MonitoringDashboard](https://github.com/kiali/kiali/blob/v1.5/operator/deploy/crd.yaml#L2-L17)) must now have a definition of a structural schema - this is mandatory for CRDs of version `apiextensions.k8s.io/v1`

So we need to get ready to bump up our versions of our CRDs and add schema definitions to them. This is going to probably involve lots of changes across different places (think OLM and the operator manifest definitions for one example).

For more, read these:

* How to migrate an operator to a new CRD: https://github.com/operator-framework/operator-sdk/blob/0481703cbf7ae25b0862f14ff4a4f5b652632897/doc/user/migrating-existing-apis.md
* Upgrading CRDs within OLM: https://github.com/operator-framework/operator-lifecycle-manager/blob/0.12.0/doc/design/dependency-resolution.md#customresourcedefinition-crd-upgrade
* How OLM upgrades work: https://github.com/operator-framework/operator-lifecycle-manager/blob/0.12.0/doc/design/how-to-update-operators.md
* How to specify a structural schema in a CRD: https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/#specifying-a-structural-schema

**Comment by jotak** [comment]
Ok, for monitoringdashboard it was kept schema-less which I think helped quite a lot at the beginning to let time for the API to stabilize. Now it's time to change.

**Comment by jotak** [comment]
For the record, I'm putting this as a gist but haven't had the chance to test and verify yet as I need to upgrade my k8s : https://gist.github.com/jotak/e335ecc6b19229b1920c55897195d1a4

**Comment by lucasponce** [comment]
I'd removed from Sprint 32 as it's waiting external.
It can be reviewed and added in Sprint 32 if it gets unblocked and it can progress.

**Comment by jotak** [comment]
@israel-hdez @jmazzitelli is this task only for Kiali CR or also for monitoring dashboards CR? (put differently, should I create a separate task?)

**Comment by jmazzitelli** [comment]
@jotak I think we should do it all together. After all, the point of this whole thing is because "The old v1beta1 is deprecated" so both of our CRDs need to be done.

**Comment by jotak** [comment]
@jmazzitelli it's more about organizational stuff: is @israel-hdez working on dashboards CRD too? Else I can create a task and do that myself in the same sprint

**Comment by jmazzitelli** [comment]
> @israel-hdez working on dashboards CRD too?

No. He has only been working on the Kiali CRD here: https://github.com/kiali/kiali/pull/1920/files#diff-cc2a416f94f07d10da981df43a05710a

**Comment by israel-hdez** [comment]
@jotak I confirm that I haven't worked on the dashboards CRDs. I was going to do it incrementally on two or three PRs and link the PRs to this same task. But if you want and since you have better knowledge of that part, you can go ahead an do the dashboard CRDs.

**Comment by israel-hdez** [comment]
This will be on-hold until OS supports the k8s 1.16 API (or until we find a safe way to implement this without maintaining two versions of the CRDs).

**Comment by jmazzitelli** [comment]
To: Kubernetes Dev <kubernetes-dev@googlegroups.com>, <kubebuilder@googlegroups.com>, K8s API Machinery SIG <kubernetes-sig-api-machinery@googlegroups.com>, <kubernetes+announcements@discoursemail.com>


The v1beta1 version of CustomResourceDefinition, MutatingWebhookConfiguration, and ValidatingWebhookConfiguration are being removed in 1.22 after six releases of deprecation.  In 1.22, the kube-apiserver will no longer serve the v1beta1 version of these resources and the only way to access them will be through the v1 APIs, available since 1.16.


We’re sending this note as an extra reminder, because these are important extensibility APIs, and even though there has been ample warning, it is in keeping with our deprecation policy, and the (better) v1 APIs have been available since 1.16, we don’t want anyone to be surprised.

All supported versions of kubernetes have the CRD, MutatingWebhookConfiguration, and ValidatingWebhookConfiguration  v1 APIs.

The PRs promoting the APIs to v1 included a deprecation notice.

Since 1.19, every request has returned a deprecation warning to the client.

Also since 1.19, the deprecation metrics have been set.

CustomResource defaulting requires CRD v1.

Given the above, we think that clients who have not yet switched to the v1 version of these APIs are unlikely to switch no matter how much additional time they are given. (But please recommend additional forums for announcing this if you can think of a way of reaching more people!)

 

If this is the first you’ve heard of this and you haven’t moved to the v1 versions of these APIs yet, a summary of the changes can be found in the 1.16 release notes and CRD structural schemas documentation.

**Comment by jmazzitelli** [comment]
We should move to the new version but allow for any/all yaml. We aren't ready yet to support actual schemas - we add so many settings (almost at least a new setting per release) this will be unwieldy very fast if we support full schema. So we should move the version and use something like `spec.preserveUnknownFields: true` or `x-kubernetes-preserve-unknown-fields: true`
